### PR TITLE
Jd/optional pj

### LINKED ISSUE #112: Map PowerModels variables into CanonicalModel

### LINKED ISSUE #113: Update initial conditions from solution

### LINKED ISSUE #114: Make use of ParameterJuMP an option
as per @claytonpbarrows comment in #111

### LINKED ISSUE #115: non-sequential bus numbers breaks canonical model construction
If I change the bus number in a 5-bus case to ... say 8, I get the following 

```
PS.EconomicDispatch(sys5b_uc,PowerSimulations.CopperPlatePowerModel,optimizer = solver)

BoundsError: attempt to access 5×24 Array{ParameterJuMP.DoubleGenericAffExpr{Float64,VariableRef,ParameterJuMP.Parameter},2} at index [8, 1]

Stacktrace:
 [1] setindex! at ./array.jl:769 [inlined]
 [2] _add_to_expression!(::Array{ParameterJuMP.DoubleGenericAffExpr{Float64,VariableRef,ParameterJuMP.Parameter},2}, ::Int64, ::Int64, ::VariableRef) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/devices/device_models/common/undef_check.jl:46
 [3] add_variable(::PowerSimulations.CanonicalModel, ::Array{ThermalDispatch,1}, ::UnitRange{Int64}, ::Symbol, ::Bool, ::Symbol) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/devices/device_models/common/add_variable.jl:36
 [4] activepower_variables(::PowerSimulations.CanonicalModel, ::Array{ThermalDispatch,1}, ::UnitRange{Int64}) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/devices/device_models/thermal_generation.jl:24
 [5] #construct_device!#123(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::PowerSimulations.CanonicalModel, ::Type{ThermalGen}, ::Type{PowerSimulations.ThermalDispatch}, ::Type{CopperPlatePowerModel}, ::PowerSystem{PowerLoad,Array{Line,1},Array{Storage,1}}, ::UnitRange{Int64}) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/devices/device_constructors/thermalgeneration_constructor.jl:211
 [6] construct_device! at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/devices/device_constructors/thermalgeneration_constructor.jl:211 [inlined]
 [7] #construct_device!#4 at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/core/device_constructor.jl:8 [inlined]
 [8] construct_device!(::PowerSimulations.CanonicalModel, ::PowerSimulations.DeviceModel{ThermalGen,PowerSimulations.ThermalDispatch}, ::Type{CopperPlatePowerModel}, ::PowerSystem{PowerLoad,Array{Line,1},Array{Storage,1}}, ::UnitRange{Int64}) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/core/device_constructor.jl:8
 [9] #build_canonical_model#7(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Type{CopperPlatePowerModel}, ::Dict{Symbol,PowerSimulations.DeviceModel}, ::Dict{Symbol,PowerSimulations.DeviceModel}, ::Dict{Symbol,PowerSimulations.ServiceModel}, ::PowerSystem{PowerLoad,Array{Line,1},Array{Storage,1}}, ::OptimizerFactory) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/core/canonical_constructor.jl:60
 [10] build_canonical_model at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/core/canonical_constructor.jl:53 [inlined]
 [11] #PowerOperationModel#2 at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/core/core_structs/operation_model.jl:24 [inlined]
 [12] Type at ./none:0 [inlined]
 [13] #EconomicDispatch#174(::OptimizerFactory, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Type, ::PowerSystem{PowerLoad,Array{Line,1},Array{Storage,1}}, ::Type{CopperPlatePowerModel}) at /Users/cbarrows/.julia/packages/PowerSimulations/fP3qG/src/operation_models/economic_dispatch.jl:13
 [14] (::getfield(Core, Symbol("#kw#Type")))(::NamedTuple{(:optimizer,),Tuple{OptimizerFactory}}, ::Type{EconomicDispatch}, ::PowerSystem{PowerLoad,Array{Line,1},Array{Storage,1}}, ::Type{CopperPlatePowerModel}) at ./none:0
 [15] top-level scope at In[38]:1
```

### LINKED ISSUE #117: remove double loops in time series constraints

### LINKED ISSUE #119: add PhaseShiftingTransfomer to PM translator
need a translator for PhaseShiftingTransformers: https://github.com/NREL/PowerSimulations.jl/blob/master/src/devices/device_models/common/pm_translator.jl

### LINKED ISSUE #123: add bridge to PM for DCLine
```
ED = PSI.EconomicDispatch(sys_rts, PSI.PM.DCPlosslessForm; optimizer = GLPK_optimizer, parameters = false);
ERROR: MethodError: no method matching get_branch_to_pm(::Int64, ::VSCDCLine)
Stacktrace:
 [1] get_branches_to_pm(::Array{Branch,1}) at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/devices/device_models/common/pm_translator.jl:102
 [2] pass_to_pm(::PowerSystem{ElectricLoad,Array{Branch,1},Array{Storage,1}}) at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/devices/device_models/common/pm_translator.jl:133
 [3] powermodels_network!(::PowerSimulations.CanonicalModel, ::Type{PowerModels.DCPlosslessForm}, ::PowerSystem{ElectricLoad,Array{Branch,1},Array{Storage,1}}, ::UnitRange{Int64}) at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/network/network_models/powermodels_interface.jl:282
 [4] #construct_network!#157 at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/network/network_constructor.jl:41 [inlined]
 [5] #construct_network! at ./none:0 [inlined]
 [6] #build_canonical_model#8(::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:parameters,),Tuple{Bool}}}, ::Function, ::Type{PowerModels.DCPlosslessForm}, ::Dict{Symbol,PowerSimulations.DeviceModel}, ::Dict{Symbol,PowerSimulations.DeviceModel}, ::Dict{Symbol,PowerSimulations.ServiceModel}, ::PowerSystem{ElectricLoad,Array{Branch,1},Array{Storage,1}}, ::OptimizerFactory) at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/core/canonical_constructor.jl:105
 [7] #build_canonical_model at ./none:0 [inlined]
 [8] #PowerOperationModel#2 at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/core/core_structs/operation_model.jl:24 [inlined]
 [9] Type at ./none:0 [inlined]
 [10] #EconomicDispatch#159(::OptimizerFactory, ::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:parameters,),Tuple{Bool}}}, ::Type, ::PowerSystem{ElectricLoad,Array{Branch,1},Array{Storage,1}}, ::Type{PowerModels.DCPlosslessForm}) at /Users/cbarrows/.julia/packages/PowerSimulations/m36VR/src/operation_models/economic_dispatch.jl:13
 [11] (::getfield(Core, Symbol("#kw#Type")))(::NamedTuple{(:optimizer, :parameters),Tuple{OptimizerFactory,Bool}}, ::Type{EconomicDispatch}, ::PowerSystem{ElectricLoad,Array{Branch,1},Array{Storage,1}}, ::Type{PowerModels.DCPlosslessForm}) at ./none:0
 [12] top-level scope at util.jl:156
```

### PR DESCRIPTION
**Pull request by jd-lara**
closes #112 
closes #113 
closes #114
closes #115 
closes #117 @daniel-thom raised concerns about this issue. I think I caught them all. 
closes #119 
closes #123 

@jmaack24 this should allow using StructJuMP and avoid the headaches related to ParameterJuMP too.

### CODE REVIEW

File: src/core/canonical_constructor.jl [hunk 1]
```diff
@@ -1,40 +1,81 @@
 
-function _canonical_model_init(bus_count::Int64,
+function _container_spec(V::DataType, ax...; kwargs...)
+
+    if :parameters in keys(kwargs)
```
Comment (daniel-thom) [inline src/core/canonical_constructor.jl]: Suggest using this one-liner in all locations:
parameters = get(kwargs, :parameters, true)

File: src/core/canonical_constructor.jl [hunk 2]
```diff
@@ -1,40 +1,81 @@
...
+        parameters = kwargs[:parameters]
+    else
+        parameters = true
+    end
+
+    # While JuMP fixes the isassigned problems
+    if parameters
```
Comment (daniel-thom) [inline src/core/canonical_constructor.jl]: IMHO, this is a perfect place for using a ternary operator.

cont = parameters ? JuMP.Containers.DenseAxisArray{PGAE{V}}(undef, ax...) :
                    JuMP.Containers.DenseAxisArray{GAE{V}}(undef, ax...)
_remove_undef!(cont.data)
return cont

File: src/core/core_structs/device_model.jl [hunk 3]
```diff
@@ -6,7 +6,7 @@ mutable struct DeviceModel{D <: PSY.PowerSystemDevice,
     formulation::Type{B}
 end
 
-mutable struct InitialCondition
+mutable struct InitialCondition{T <: Union{PJ.Parameter, Float64}}
     device::PSY.PowerSystemDevice
-    value::Union{PJ.Parameter, Float64}
+    value::T
 end
```
Comment (daniel-thom) [inline src/core/core_structs/device_model.jl]: You should add a newline to the file.

File: src/core/core_structs/operation_model.jl [hunk 4]
```diff
@@ -20,7 +20,7 @@ mutable struct PowerOperationModel{M <: AbstractOperationsModel,
                                 optimizer::Union{Nothing,JuMP.OptimizerFactory}=nothing,
                                 kwargs...) where {M <: AbstractOperationsModel,
                                                   T <: PM.AbstractPowerFormulation}
-
+
```
Comment (daniel-thom) [inline src/core/core_structs/operation_model.jl]: nit: extra whitespace

File: src/core/operations_constructor.jl [hunk 5]
```diff
@@ -17,14 +23,14 @@ function _pass_abstract_jump(optimizer::Union{Nothing,JuMP.OptimizerFactory}; kw
     end
 
     JuMPmodel = JuMP.Model(optimizer)
-    PJ.enable_parameters(JuMPmodel)
+    parameters ? PJ.enable_parameters(JuMPmodel) : return JuMPmodel
```
Comment (daniel-thom) [inline src/core/operations_constructor.jl]: Here I would not use a ternary operator.

if parameters
    PJ.enable_parameters(JuMPmodel)
end

return JuMPmodel

File: src/core/canonical_constructor.jl [hunk 6]
```diff
@@ -1,40 +1,81 @@
...
                             Dict{Symbol, JuMP.Containers.DenseAxisArray}(),
                             Dict{Symbol, JuMP.Containers.DenseAxisArray}(),
-                            zero(JuMP.GenericAffExpr{Float64, JuMP.variable_type(jump_model)}),
-                            Dict{Symbol, JuMPPGAEArray{V}}(:var_active => JuMPPGAEArray{V}(undef, bus_count, time_periods),
-                                                                        :var_reactive => JuMPPGAEArray{V}(undef, bus_count, time_periods)),
-                            Dict{Symbol,Any}(),
+                            zero(JuMP.GenericAffExpr{Float64, V}),
+                            Dict{Symbol, JuMP.Containers.DenseAxisArray}(:var_active => _container_spec(V, bus_numbers, time_range; kwargs...),
+                                                                         :var_reactive => _container_spec(V, bus_numbers, time_range; kwargs...)),
+                            parameters ? Dict{Symbol,JuMP.Containers.DenseAxisArray}() : nothing,
```
Comment (daniel-thom) [inline src/core/canonical_constructor.jl]: It seems like you could avoid passing "parameters" throughout the stack like you're doing.  It's already implicitly part of CanonicalModel.  You just need to check the type of ps_m.parameters, which you could put in a helper function.

Also, I don't see any documentation for callers about what passing in "parameters" is going to do.

File: src/devices/device_models/thermal_generation.jl [hunk 7]
```diff
@@ -245,19 +247,16 @@ function ramp_constraints(ps_m::CanonicalModel,
     if !isempty(data[2])
```
Comment (daniel-thom) [inline src/devices/device_models/thermal_generation.jl]: This isn't new to this PR, but you have a bit of duplicated code in these ramp_constraints functions (it still is a recent change).  You could consolidate the logic in a function.  The only difference is the function device* that gets called. You could pass that in to the new function.

File: src/core/canonical_constructor.jl (hunk 2 above)
Comment (jd-lara) [reply src/core/canonical_constructor.jl]: This is a temporary code while a fix is added to JuMP to address the `isassigned` function. The code will be deleted later.

File: src/core/canonical_constructor.jl (hunk 6 above)
Comment (jd-lara) [reply src/core/canonical_constructor.jl]: I think is only applicable sometimes depending on when the parameters are created. Will leave it for a future PR to check where is good to use.

**Pull request merged**